docs: add Dialog header/footer sections (#1307)
vaadin/docs: articles/ds/components/dialog/index.asciidoc

  - 'Source: https://github.com/vaadin/web-components/tree/v{moduleNpmVersion:vaadin-dialog}/packages/vaadin-dialog[TypeScript] / https://github.com/vaadin/flow-components/tree/{moduleMavenVersion:com.vaadin:vaadin}/vaadin-dialog-flow-parent[Java]'
---

> :experimental:
> 
> = Dialog

Reviewer: Never put any content/attributes/config before the main page heading. That messes up the placement of the TOC in the website. Might be worth a lint error…

```suggestion
= Dialog
:experimental:
```

</style>
++++

== Structure
The Dialog component has static header and footer areas, and a scrolling content area in between.
The header and footer are optional, and automatically hidden if empty and not explicitly enabled.

=== Header
The header contains an optional title element, and a slot next to it for custom header content, such as a close button.

[.example]
--
include::../_shared.asciidoc[tag=guard-directive-note]

[source,html]
----
include::{root}/frontend/demo/component/dialog/dialog-header.ts[render,tags=*,indent=0,group=TypeScript]
----

[source,java]
----
include::{root}/src/main/java/com/vaadin/demo/component/dialog/DialogHeader.java[render,tags=*,indent=0,group=Java]
----
--

=== Footer

> Buttons for closure actions such as _Save_, _Cancel_, _Delete_ etc should be placed in the footer. 

Reviewer 2: ```suggestion
Buttons for closure actions, such as _Save_, _Cancel_, _Delete_, etc., should be placed in the footer. 
```

See <<../button#buttons-in-dialogs,Button>> component for guidelines for the placement of buttons in dialogs. Footer content is right-aligned by default.
Components can be left-aligned by applying a margin:

[.example]
--
include::../_shared.asciidoc[tag=guard-directive-note]

[source,html]
----
include::{root}/frontend/demo/component/dialog/dialog-footer.ts[render,tags=*,indent=0,group=TypeScript]
----

[source,java]
----
include::{root}/src/main/java/com/vaadin/demo/component/dialog/DialogFooter.java[render,tags=*,indent=0,group=Java]
----
--

=== Padding

The content area's built-in padding can be removed by applying the `no-padding` theme variant.

[.example]
--
include::../_shared.asciidoc[tag=guard-directive-note]

[source,html]
----
include::{root}/frontend/demo/component/dialog/dialog-no-padding.ts[render,tags=*,indent=0,group=TypeScript]
----

[source,java]
----
include::{root}/src/main/java/com/vaadin/demo/component/dialog/DialogNoPadding.java[render,tags=snippet,indent=0,group=Java]
----

[source,css]
----
include::{root}/frontend/themes/docs/dialog.css[tags=dialog-no-padding,indent=0,group=Java]

> ----

Reviewer: This file is unnecessary now, isn't it?

Author: Good catch! Removed.

Any component contained within a Dialog can be marked and used as a drag handle by applying the `draggable` class name to it.
You can choose whether or not to make the component's content draggable as well, or just the component itself.

It is recommended to make non-modal Dialogs draggable so that the user can interact with content that might otherwise be obscured by the Dialog.
For example, a Dialog for taking notes, or adding widgets to a dashboard using drag and drop, can offer a better experience by allowing the user to move the Dialog around.

> Modal Dialogs don't benefit from being draggable as their modality curtain (the dark overlay behind the dialog) obscures the underlying user interface.

Reviewer 2: ```suggestion
Modal Dialogs do not benefit from being draggable, as their modality curtain (the dark overlay behind the dialog) obscures the underlying user interface.
```

Any component contained within a Dialog can be marked and used as a drag handle by applying the `draggable` class name to it.
You can choose whether or not to make the component's content draggable as well, or just the component itself.

> It is recommended to make non-modal Dialogs draggable so that the user can interact with content that might otherwise be obscured by the Dialog.

Author: There's one item here under discussion on the draft doc that was not added yet.

Reviewer 3: Which is that?

Author: It's not exactly here, but before the paragraph which starts with "Any component...". to add this:

> The default areas from which a Dialog can be dragged depend on whether the built-in header is used:
> - If the built-in header or footer is used, they function as the Dialog’s default drag handles
> - Without the built-in header, any empty space within the dialog functions as a drag handle

Reviewer 3: I think it's good enough as it is now. But I think I would move the examples down a bit, just before the paragraph starting with "It is recommended to make non-modal Dialogs draggable..."

Author: Done.
